### PR TITLE
Add support for MPSProfiler

### PR DESCRIPTION
**Pull request by razarmehr**
- Enable event and interval-based os signpost tracing
- Enable logging of MPS graphs, native kernels, and copies and their GPU times
- Enable dumping the table of kernel profiling results sorted based on Mean GPU time when the process ends.
- Fix a bug in MPSAllocator where the Allocator completionHandlers were called after MPSAllocator instance was destroyed (could also fix #82702).

### CODE REVIEW

File: aten/src/ATen/mps/MPSProfiler.mm [hunk 1]
```diff
@@ -0,0 +1,306 @@
...
+void MPSProfiler::addProfilerCompletedHandler(Profiler::BaseInfo* info, SignpostTypes signpostType, SyncType syncType) {
+  const uint64_t signpostId = info->signpostId;
+  // signpost ID is used only for interval-based signposts, and must be non-zero
+  if (m_profile_options & ProfileOptions::USE_INTERVALS) {
+    TORCH_INTERNAL_ASSERT_DEBUG_ONLY(signpostId, "Signpost interval has no identifier!");
+  }
+  info->signpostId = 0; // reset signpostId for sanity check on next interval
+
+  auto m_stream = getDefaultMPSStream();
+  // NOTE: the following block isn't thread-safe, so make sure to capture profiler's info locally
```
Comment (ssaladis) [inline aten/src/ATen/mps/MPSProfiler.mm:161]: This looks correct to me. Sorry if I mis-understood yesterday. I thought you weren't using these instances

Comment (ssaladis) [reply aten/src/ATen/mps/MPSProfiler.mm:161]: Was the issue with gpuTime discrepancy discussed yesterday, solved ?

Comment (razarmehr) [reply aten/src/ATen/mps/MPSProfiler.mm:161]: We need to add scheduleHandlers to get accurate time for GPU of ops with signposts. It should resolve the tracing intervals issues when visualizing on tools.

**Review by kulinseth — APPROVED** [review]
Looks good.

**Pull request merged**